Fix for issue #2872

iokomin opened the pull request:
#### Description:

Implemented UEFI boot mode OVAL check based on '/sys/firmware/efi' existence (assumed to be supported on most of Linux based systems). Extended grub2 ownership/permissions rules to use UEFI OVAL check: pass non-EFI checks on EFI systems and pass EFI checks on non-EFI systems.

#### Rationale:

- Fixes #2872

#### Testing:
- Checked build to pass for all targets
- Checked rules to work on OL7

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/ComplianceAsCode/content/inspections/5eab98f3-fad7-42f8-992d-dc336963269d) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/ComplianceAsCode/content/inspections/5eab98f3-fad7-42f8-992d-dc336963269d) completed: **1 new issues**

redhatrises commented on linux_os/guide/system/bootloader-grub2/file_groupowner_efi_grub2_cfg/oval/shared.xml:
```diff
@@ -0,0 +1,26 @@
+<def-group>
+  <definition class="compliance" id="file_groupowner_efi_grub2_cfg" version="1">
+    <metadata>
+      <title>Verify /boot/efi/EFI/redhat/grub.cfg Group Owner</title>
```
Should be `Verify the UEFI Boot Loader grub.cfg Group Owner`

redhatrises commented on linux_os/guide/system/bootloader-grub2/file_groupowner_efi_grub2_cfg/oval/shared.xml:
```diff
@@ -0,0 +1,26 @@
+  </unix:file_state>
+  <unix:file_object comment="/boot/efi/EFI/redhat/grub.cfg" id="object_file_groupowner_efi_grub2_cfg" version="1">
+    <unix:path>/boot/efi/EFI/redhat</unix:path>
+    <unix:filename>grub.cfg</unix:filename>
```
Collapse lines 23-24 to:
```<ind:filepath operation="pattern match">^/boot/efi/EFI/(redhat|fedora)/grub.cfg$</ind:filepath>```

redhatrises commented on linux_os/guide/system/bootloader-grub2/file_groupowner_grub2_cfg/oval/shared.xml:
```diff
@@ -0,0 +1,26 @@
+  </unix:file_state>
+  <unix:file_object comment="/boot/grub2/grub.cfg" id="object_file_groupowner_grub2_cfg" version="1">
+    <unix:path>/boot/grub2</unix:path>
+    <unix:filename>grub.cfg</unix:filename>
```
Collapse lines 23-24 to:
```<ind:filepath operation="pattern match">^/boot/efi/EFI/(redhat|fedora)/grub.cfg$</ind:filepath>```

redhatrises commented on linux_os/guide/system/bootloader-grub2/file_owner_efi_grub2_cfg/oval/shared.xml:
```diff
@@ -0,0 +1,26 @@
+  </unix:file_state>
+  <unix:file_object comment="/boot/efi/EFI/redhat/grub.cfg" id="object_file_owner_efi_grub2_cfg" version="1">
+    <unix:path>/boot/efi/EFI/redhat</unix:path>
+    <unix:filename>grub.cfg</unix:filename>
```
Collapse lines 23-24 to:
```<ind:filepath operation="pattern match">^/boot/efi/EFI/(redhat|fedora)/grub.cfg$</ind:filepath>```

redhatrises commented on linux_os/guide/system/bootloader-grub2/file_owner_grub2_cfg/oval/shared.xml:
```diff
@@ -0,0 +1,26 @@
+  </unix:file_state>
+  <unix:file_object comment="/boot/grub2/grub.cfg" id="object_file_owner_grub2_cfg" version="1">
+    <unix:path>/boot/grub2</unix:path>
+    <unix:filename>grub.cfg</unix:filename>
```
Collapse lines 23-24 to:
```<ind:filepath>/boot/grub2/grub.cfg</ind:filepath>```

redhatrises commented on linux_os/guide/system/bootloader-grub2/file_permissions_grub2_cfg/oval/shared.xml:
```diff
@@ -0,0 +1,40 @@
+  </unix:file_state>
+  <unix:file_object comment="/boot/grub2/grub.cfg" id="object_file_permissions_grub2_cfg" version="1">
+    <unix:path>/boot/grub2</unix:path>
+    <unix:filename>grub.cfg</unix:filename>
```
Collapse lines 23-24 to:
```<ind:filepath>/boot/grub2/grub.cfg</ind:filepath>```

redhatrises commented on shared/checks/oval/system_boot_mode_is_uefi.xml:
```diff
@@ -0,0 +1,24 @@
+
+  <unix:file_object comment="/sys/firmware/efi" id="object_file_sys_firmware_efi" version="1">
+    <unix:path>/sys/firmware/efi</unix:path>
+    <unix:filename xsi:nil="true" />
```
Collapse lines 23-24 to:
```<ind:filepath>/sys/firmware/efi</ind:filepath>```

iokomin replied:
/sys/firmware/efi is a directory and according to OVAL spec directory can't be specified as a filepath:
https://oval.mitre.org/language/version5.11/ovaldefinition/documentation/unix-definitions-schema.html#file_object
@redhatrises is suggested change still relevant?

redhatrises commented on linux_os/guide/system/bootloader-grub2/file_groupowner_grub2_cfg/oval/shared.xml:
```diff
@@ -0,0 +1,26 @@
+      <affected family="unix">
+        <platform>multi_platform_all</platform>
+      </affected>
+      <description>This test makes sure that /boot/grub2/grub.cfg is group owned by 0.</description>
```
`Verify that /boot/grub2/grub.cfg is group owned by gid 0.`

redhatrises commented on linux_os/guide/system/bootloader-grub2/file_owner_grub2_cfg/oval/shared.xml:
```diff
@@ -0,0 +1,26 @@
+      <affected family="unix">
+        <platform>multi_platform_all</platform>
+      </affected>
+      <description>This test makes sure that /boot/grub2/grub.cfg is owned by 0.</description>
```
`Verify that /boot/grub2/grub.cfg is owned by uid 0.`

redhatrises commented on linux_os/guide/system/bootloader-grub2/file_permissions_grub2_cfg/oval/shared.xml:
```diff
@@ -0,0 +1,40 @@
+        <platform>multi_platform_all</platform>
+      </affected>
+      <description>This test makes sure that /boot/grub2/grub.cfg has mode 0600.
+      If the target file or directory has an extended ACL, then it will fail the mode check.
```
Should be `File permissions for /boot/grub2/grub.cfg should be set to 0600 (or stronger).`

redhatrises commented on linux_os/guide/system/bootloader-grub2/file_owner_efi_grub2_cfg/oval/shared.xml:
```diff
@@ -0,0 +1,26 @@
+      <affected family="unix">
+        <platform>multi_platform_all</platform>
+      </affected>
+      <description>This test makes sure that /boot/efi/EFI/redhat/grub.cfg is owned by 0.</description>
```
`Verify that EFI bootloader grub.cfg is owned by uid 0.`

redhatrises commented on linux_os/guide/system/bootloader-grub2/file_owner_efi_grub2_cfg/oval/shared.xml:
```diff
@@ -0,0 +1,26 @@
+<def-group>
+  <definition class="compliance" id="file_owner_efi_grub2_cfg" version="1">
+    <metadata>
+      <title>Verify /boot/efi/EFI/redhat/grub.cfg Owner</title>
```
`Verify the UEFI Boot Loader grub.cfg Owner`

redhatrises commented on linux_os/guide/system/bootloader-grub2/file_owner_efi_grub2_cfg/oval/shared.xml:
```diff
@@ -0,0 +1,26 @@
+      <description>This test makes sure that /boot/efi/EFI/redhat/grub.cfg is owned by 0.</description>
+    </metadata>
+    <criteria operator="OR">
+      <extend_definition definition_ref="system_boot_mode_is_uefi" negate="true" comment="Pass if system boot mode is legacy BIOS" />
```
s/legacy BIOS/not UEFI/g

redhatrises commented on linux_os/guide/system/bootloader-grub2/file_permissions_efi_grub2_cfg/oval/shared.xml:
```diff
@@ -10,7 +10,8 @@
     </metadata>
-    <criteria>
+    <criteria operator="OR">
+      <extend_definition definition_ref="system_boot_mode_is_uefi" negate="true" comment="Pass if system boot mode is legacy BIOS" />
```
s/legacy BIOS/not UEFI/g

redhatrises commented on linux_os/guide/system/bootloader-grub2/file_permissions_grub2_cfg/oval/shared.xml:
```diff
@@ -0,0 +1,40 @@
+      <criterion comment="Check file mode of /boot/grub2/grub.cfg" test_ref="test_file_permissions_grub2_cfg" />
+    </criteria>
+  </definition>
+  <unix:file_test check="all" check_existence="all_exist" comment="Testing mode of /boot/grub2/grub.cfg" id="test_file_permissions_grub2_cfg" version="1">
```
s/Testing/Verify/g

redhatrises commented on shared/checks/oval/system_boot_mode_is_uefi.xml:
```diff
@@ -0,0 +1,24 @@
+    </criteria>
+  </definition>
+
+  <unix:file_test check="all" check_existence="at_least_one_exists" comment="Test if /sys/firmware/efi folder exists" id="test_efi_dir_existence" version="1">
```
s/Test if/Verify if/g

redhatrises commented on linux_os/guide/system/bootloader-grub2/file_owner_efi_grub2_cfg/oval/shared.xml:
```diff
@@ -0,0 +1,26 @@
+      <criterion comment="Check file ownership of /boot/efi/EFI/redhat/grub.cfg" test_ref="test_file_owner_efi_grub2_cfg" />
+    </criteria>
+  </definition>
+  <unix:file_test check="all" check_existence="all_exist" comment="Testing user ownership of /boot/efi/EFI/redhat/grub.cfg" id="test_file_owner_efi_grub2_cfg" version="1">
```
s/Testing/Verify/g

redhatrises commented on linux_os/guide/system/bootloader-grub2/file_groupowner_grub2_cfg/oval/shared.xml:
```diff
@@ -0,0 +1,26 @@
+      <criterion comment="Check file group ownership of /boot/grub2/grub.cfg" test_ref="test_file_groupowner_grub2_cfg" />
+    </criteria>
+  </definition>
+  <unix:file_test check="all" check_existence="all_exist" comment="Testing group ownership of /boot/grub2/grub.cfg" id="test_file_groupowner_grub2_cfg" version="1">
```
s/Testing/Verify/g

redhatrises commented on linux_os/guide/system/bootloader-grub2/file_groupowner_efi_grub2_cfg/oval/shared.xml:
```diff
@@ -0,0 +1,26 @@
+      <criterion comment="Check file group ownership of /boot/efi/EFI/redhat/grub.cfg" test_ref="test_file_groupowner_efi_grub2_cfg" />
+    </criteria>
+  </definition>
+  <unix:file_test check="all" check_existence="all_exist" comment="Testing group ownership of /boot/efi/EFI/redhat/grub.cfg" id="test_file_groupowner_efi_grub2_cfg" version="1">
```
s/Testing/Verify/g

redhatrises commented on linux_os/guide/system/bootloader-grub2/file_groupowner_efi_grub2_cfg/oval/shared.xml:
```diff
@@ -0,0 +1,26 @@
+      <affected family="unix">
+        <platform>multi_platform_all</platform>
+      </affected>
+      <description>This test makes sure that /boot/efi/EFI/redhat/grub.cfg is group owned by 0.</description>
```
should be `Verify that UEFI grub.cfg is group owned by gid 0.`

redhatrises commented:
@iokomin the update the template CSV files as well.

iokomin commented:
@redhatrises could you please clarify what change to CSV files expected? 
Note: grub.cfg entries in CSV files would allow to generate bash and ansible remediation scripts while OVAL content will be taken from linux_os.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/ComplianceAsCode/content/inspections/e9ad4b47-a8f0-496b-8d1e-cb13b336a8a5) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/ComplianceAsCode/content/inspections/e9ad4b47-a8f0-496b-8d1e-cb13b336a8a5) completed: **No new issues**

iokomin commented:
Thanks for taking your time.
Added fixes according to review conversations, branch rebased, please review.

iokomin commented:
Guys, could please anyone check updated PR? Is there anything else to be fixed/changed in suggested solution?

shawndwells commented:
On 4/8/19 5:23 AM, Ilya Okomin wrote:
> Guys, could please anyone check updated PR? Is there anything else to 
> be fixed/changed in suggested solution?


@redhatrises was out of office last week. Should be able to (re)review 
later today. I'll keep an eye on it too.

redhatrises commented:
@iokomin sorry for the wait... Please make sure to enable the rules for fedora as well.

Tested this, and I don't think the logic is right. For example, on a non-EFI system:

`
Title   Verify /boot/grub2/grub.cfg Group Ownership
Rule    xccdf_org.ssgproject.content_rule_file_groupowner_grub2_cfg
Result  pass

Title   Verify /boot/grub2/grub.cfg User Ownership
Rule    xccdf_org.ssgproject.content_rule_file_owner_grub2_cfg
Result  pass

Title   Verify /boot/efi/EFI/redhat/grub.cfg Group Ownership
Rule    xccdf_org.ssgproject.content_rule_file_groupowner_efi_grub2_cfg
Result  fail

Title   Verify /boot/efi/EFI/redhat/grub.cfg User Ownership
Rule    xccdf_org.ssgproject.content_rule_file_owner_efi_grub2_cfg
Result  fail
`
The EFI checks should pass, but they don't.

redhatrises commented:
> @iokomin sorry for the wait... Please make sure to enable the rules for fedora as well.
> 
> Tested this, and I don't think the logic is right. For example, on a non-EFI system:
> 
> ```
> Title   Verify /boot/grub2/grub.cfg Group Ownership
> Rule    xccdf_org.ssgproject.content_rule_file_groupowner_grub2_cfg
> Result  pass
> 
> Title   Verify /boot/grub2/grub.cfg User Ownership
> Rule    xccdf_org.ssgproject.content_rule_file_owner_grub2_cfg
> Result  pass
> 
> Title   Verify /boot/efi/EFI/redhat/grub.cfg Group Ownership
> Rule    xccdf_org.ssgproject.content_rule_file_groupowner_efi_grub2_cfg
> Result  fail
> 
> Title   Verify /boot/efi/EFI/redhat/grub.cfg User Ownership
> Rule    xccdf_org.ssgproject.content_rule_file_owner_efi_grub2_cfg
> Result  fail
> ```
> 
> The EFI checks should pass, but they don't.

Nevermind. Retested on a clean branch, and it worked.

@iokomin only thing left is to enable `fedora` in the rules.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/ComplianceAsCode/content/inspections/e427398d-3603-416a-b14e-677bdd40d6bb) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/ComplianceAsCode/content/inspections/e427398d-3603-416a-b14e-677bdd40d6bb) completed: **2 new issues**

iokomin commented:
@redhatrises please find additional commit with fedora support added:
- added fedora prodtype
- updated rules  description to provide correct grub.cfg path for fedora, because of this updated titles to be more generic
- fixed file_permissions_efi_grub2_cfg OVAL to be applicable on fedora OS

Checked generated content for fedora,rhel7,ol7 builds. Checked boot loader permission/ownership rules to work on OL7.

redhatrises commented on linux_os/guide/system/bootloader-grub2/file_groupowner_efi_grub2_cfg/rule.yml:
```diff
@@ -30,6 +37,14 @@ references:
+ocil: '{{{ ocil_file_group_owner(file="/boot/efi/EFI/fedora/grub.cfg", group="root") }}}'
+{{% else %}}
 ocil: '{{{ ocil_file_group_owner(file="/boot/efi/EFI/redhat/grub.cfg", group="root") }}}'
+{{% endif %}}
```
@iokomin please don't wrap `ocil` or `ocil_clause` YAML keys in Jinja. Please fix.

redhatrises commented on linux_os/guide/system/bootloader-grub2/file_owner_efi_grub2_cfg/rule.yml:
```diff
@@ -28,6 +35,14 @@ references:
+ocil: '{{{ ocil_file_owner(file="/boot/efi/EFI/fedora/grub.cfg", owner="root") }}}'
+{{% else %}}
 ocil: '{{{ ocil_file_owner(file="/boot/efi/EFI/redhat/grub.cfg", owner="root") }}}'
+{{% endif %}}
```
@iokomin please don't wrap `ocil` or `ocil_clause` YAML keys in Jinja. Please fix.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/ComplianceAsCode/content/inspections/4a05e96c-0bcc-47fe-aa85-e8cc2b9b0818) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/ComplianceAsCode/content/inspections/4a05e96c-0bcc-47fe-aa85-e8cc2b9b0818) completed: **3 new issues**

iokomin commented:
@redhatrises thanks for review, your comment makes sense to me.
Please take a look at the updated patch.
Note: had to update macros.jinja - added indentation to ocil_clause_file macros to avoid "yaml.scanner.ScannerError: while scanning a simple key" for multi-line case usage in rule.yml.

redhatrises commented:
@iokomin one more thing needs to be fixed before merging. 

With the changes to `macros.jinja`, getting extra whitespace or newlines added to OCIL.

Extra whitespace added:
```
that  /boot/grub2/grub.cfg has owner root
     ^ extra space
```

Added newlines:
```
Is it the case that 

/boot/efi/EFI/redhat/grub.cfg has group owner root?
```

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/ComplianceAsCode/content/inspections/8aea00de-0432-41eb-ac19-21bd75cec88a) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/ComplianceAsCode/content/inspections/8aea00de-0432-41eb-ac19-21bd75cec88a) completed: **1 new issues**

iokomin commented:
@redhatrises please find patch to resolve extra spaces and newlines issue. 
I noticed mentioned extra spaces are also available in master. With the fix OCIL question_text items look good, hope without introduced issues this time.
Checked rules to pass on OL7, checked build and tests to pass for all targets.
Note: BTW could you please suggest how generated OCIL could/should be checked other than reviewing it's content manually?

redhatrises commented:
Thanks for all the hard work @iokomin 

LGTM.
